torch/csrc/distributed/rpc/init.cpp

```diff
@@ -41,7 +42,26 @@ PyObject* rpc_init(PyObject* /* unused */) {
           [&](FutureMessage& fut) {
             return to_py_obj(fut.wait());
           },
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+      .def("get",
+          [&](FutureMessage& fut) {
+            auto ret = c10::optional<py::object>();
+            if (fut.message().has_value()) {
+              ret = to_py_obj(fut.message().value());
+            }
+            return ret;
+          },
+          py::call_guard<py::gil_scoped_release>())
+      .def("then",
+          // Python Callback taking a FutureMessage& instead of py::object of
```

Comment:
One case would be `FutureMessage.then(log_finish_time)`. Or more generally, there could be some scenarios using RPC to synchronize distributed execution by, e.g., modifying some remote global state. This would be especially useful if we do remove APIs like sync/join.

Reply:
Could we have two variations of `.then` to address this? One in which the callback receives a `py::object` as input since the user needs the result and one which doesn't pass any parameters to the callback. We could call them `then` and `then_without_result`. I just feel that having the callback receive a Future isn't the right API since the user might think that this Future might not yet be ready (whereas its always ready). Although, I feel that for the first version we should have callback receive a py::object and if a user has needs where they don't need the result and the conversion from Message to py::object is an overhead, we could then introduce the `then_without_result` API.